### ISSUE TITLE
Dereference Spring/other function calls before script execution

### ISSUE THREAD

**Issue by moreginger** [ZeroK-RTS/Zero-K#5142]
Many / most of our Lua scripts dereference functions on the Spring table at the top of the script.

e.g. `local spGetGameFrame = Spring.GetGameFrame`

It is believed that this is faster. In fact, this is the case. For example:

```
    for i = 1, 1000000 do
        Spring.GetGameFrame()
    end
```

Takes around 0.025s, whereas calling `spGetGameFrame()` takes 0.017s.

_This is an extreme example requiring the least possible computation in the engine_. For comparison 100000 calls (10x less than the above test) to `Spring.GetVisibleUnits()` takes around 0.045s whereas the dereferenced function is not appreciably faster. If we assume the table lookup is constant then we'd expect something like 0.008 (from the first test) / 10 (ten times fewer calls) = < 0.001s difference.

Nevertheless the first case demonstrates this technique can be faster. To avoid doing the dereferencing by hand, is is possible to precompile our Lua scripts so that all Spring functions, and other common things like math, are dereferenced once in a header section?

**Comment by sprunk** [comment]
Doable, see https://github.com/ZeroK-RTS/Zero-K/blob/master/LuaUI/Widgets/chili/Headers/autolocalizer.lua

IDK if it's a good idea though. It can make things behave in ways the dev didn't expect which is bad for modding and it incurs overhead elsewhere (the `spBla` var is potential garbage which has to be checked each time by the garbage collector).

**Comment by moreginger** [comment]
> IDK if it's a good idea though. It can make things behave in ways the dev didn't expect which is bad for modding and it incurs overhead elsewhere (the `spBla` var is potential garbage which has to be checked each time by the garbage collector).

Yeah I'd be happy if we issued advice on when to do it, especially if the GC overhead is a hidden tax that is easily missed (i.e. not profiled).

**Comment by GoogleFrog** [comment]
In recent years I have begun to consider the memory cost. If a function is used infrequently, does the minuscule performance gain outweigh the minuscule memory cost?

**Comment by sprunk** [comment]
There can be a middle ground too. `local sp = Spring` and then doing `sp.Bla` helps all Spring calls a bit (because with it being a global, it's essentially indexing an extra table indexing: `_G["Spring"].Bla`) while being just one var per widget. Perhaps it's even better than that because then all functions no longer need to individually store the string "Spring" to know what to index the global table with.

**Comment by moreginger** [comment]
> There can be a middle ground too. `local sp = Spring` and then doing `sp.Bla` helps all Spring calls a bit (because with it being a global, it's essentially indexing an extra table indexing: `_G["Spring"].Bla`) while being just one var per widget. Perhaps it's even better than that because then all functions no longer need to individually store the string "Spring" to know what to index the global table with.

Looks like that's about half way, or slightly worse, between the two other options for execution time. Definitely worth considering.
